File: leetcode/1-TwoSum/TwoSum.java
/*
 * [Source]
 * 
 * https://leetcode.com/problems/two-sum/
 *
 * [Problem Description]
 *
 * Given an array of integers, return indices of the two numbers such that 
 * they add up to a specific target.
 * You may assume that each input would have exactly one solution, 
 * and you may not use the same element twice.
 * 
 * Given nums = [2, 7, 11, 15], target = 9,
 * 
 * Because nums[0] + nums[1] = 2 + 7 = 9,
 * return [0, 1].
 *
 * [Companies]
 */

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class TwoSum
{
  // We use a dictionary ADT (hash map) that contains <idx, value> pairs
  // to reduce the look up time of a value to O(1).
  public static int[] twoSum(int[] numbers, int target)
  {
    Map<Integer, Integer> dict = new HashMap<>();
    for (int i = 0; i < numbers.length; i++)
    {
      int x = numbers[i];
      if (dict.containsKey(target - x))
      {
        return new int[] {dict.get(target - x), i};
      }
      dict.put(x, i);
    }
    throw new IllegalArgumentException("No two sum solution");
  }

  public static void main(String[] args)
  {
    int[] num = {2,7,11,15};
    int target = 9;
    System.out.println(Arrays.toString(twoSum(num, target)));
  }
}

File: leetcode/1-TwoSum/TwoSum.rs
use std::collections::HashMap;

#[allow(dead_code)]
struct Solution {}

impl Solution {
    #[allow(dead_code)]
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut dict: HashMap<i32, i32> = HashMap::new();
        for idx in 0..nums.len() {
            if dict.contains_key(&(target - nums[idx])) {
                return vec![dict[&(target - nums[idx])], idx as i32];
            } else {
                dict.insert(nums[idx], idx as i32);
            }
        }
        panic!("No solution to TwoSum");
    }
}

#[test]
fn test_1() {
    let nums = vec![2, 7, 11, 15];
    let target = 9;
    assert_eq!(Solution::two_sum(nums, target), vec![0, 1]);
}
